### ISSUE TITLE
Hooking extern functions in linux kernel module (ARM)

### ISSUE THREAD

**Issue by sraboy** [angr/angr#546]
Using `Project.hook_symbol('strncmp', strncmphook)` doesn't appear to work in my kernel module. I can hook each call site individually but that obviously takes more manual analysis time. Essentially, I want this symbol to resolve to a SimProcedure:
```
In [115]: b = angr.Project("./led-kmod.ko")
     ...: x=b.loader.main_bin.get_symbol('strncmp')
     ...: print "%s is at %s" % (x.name, hex(x.addr))
     ...: 
WARNING | 2017-07-27 14:55:10,285 | cle.loader | The main binary is a position-independent executable. It is being loaded with a base address of 0x400000.
strncmp is at 0x0
```
Hooking each `BL strncmp` call site with `b.hook(addr, strncmphook, length=4)` works, where `strncmphook` is a function that manually resolves R0, R1 and R2 to compare the strings at those addresses and sets R0 appropriately.

How can I properly hook this `extern` symbol?

**Comment by sraboy** [comment]
Sorry, I was in a rush out the door when I posted yesterday. Here's some more detail:

The issue isn't specific to `strncmp` as I do the same with `mutex_lock_nested`. Here's the relevant disassembly and my full solution script. I've commented out the `strncmp` call site hooks (that work). Trying to `hook_symbol` results in the code running forever (never bothered to let it finish) on the `pg.explore(find=0x400184, avoid=0x40023c).unstash(from_stash='found',to_stash='active')` line towards the end. Ditto if I attempt to set the hook at the entry in `extern` at 0x400e94.

I'm on the latest angr-dev checkout.

(If the binary looks familiar, it's from the Gumstix Door Node from the latest DARPA VET engagement... not sure if you guys worked with the same binaries or not. I'm trying to re-solve the malicious atoms with Angr.)

```
.text:0000012C                 STMFD   SP!, {R4,R5,LR}
.text:00000130                 MOV     R4, R1
.text:00000134                 SUB     SP, SP, #0xC
.text:00000138                 MOV     R1, #0
.text:0000013C                 LDR     R0, =acmeled_mutex
.text:00000140                 MOV     R5, R2
.text:00000144                 BL      mutex_lock_nested
.text:00000148                 SUB     R12, R4, #1
.text:0000014C                 ADD     R1, R4, #6
.text:00000150                 ADD     LR, SP, #0x18+encoded
.text:00000154
.text:00000154 loc_154                                 ; CODE XREF: dev_write+3Cj
.text:00000154                 LDRB    R3, [R12,#1]!
.text:00000158                 EOR     R3, R3, #0x39
.text:0000015C                 MVN     R3, R3
.text:00000160                 CMP     R12, R1
.text:00000164                 STRB    R3, [LR],#1
.text:00000168                 BNE     loc_154
.text:0000016C                 ADD     R0, SP, #0x18+encoded ; s1
.text:00000170                 LDR     R1, =aOpen      ; "OPEN"
.text:00000174                 MOV     R2, #4          ; n
.text:00000178                 BL      strncmp
.text:0000017C                 CMP     R0, #0
.text:00000180                 BEQ     loc_23C

<snip>

extern:00000E90                 IMPORT device_create    ; CODE XREF: init_module+BCp
extern:00000E94 ; int strncmp(const char *s1, const char *s2, size_t n)
extern:00000E94                 IMPORT strncmp          ; CODE XREF: dev_write+4Cp
extern:00000E94                                         ; dev_write+64p ...
```

```
import angr, claripy

def mutexlocknestedhook(state):
    pass

'''
I wanted to replace the kernel strncmp call with libc's but I can't seem 
to actually hook the .extern entry so I rewrote it here and hooked every
call to it below.
'''
def strncmphook(state):
    print "r0: ", hex(state.regs.r0.args[0])
    print "r1: ", hex(state.regs.r1.args[0])
    print "r2: ", hex(state.regs.r2.args[0])
    print "Stack: "
    print repr(state.se.any_str(state.memory.load(state.regs.sp.args[0], 64)))
    r0str = state.se.any_str(state.memory.load(state.regs.r0.args[0], state.regs.r2.args[0]))
    r1str = state.se.any_str(state.memory.load(state.regs.r1.args[0], state.regs.r2.args[0]))
    print "Real *r0 string: ", r0str
    print "Real *r2 string: ", r1str
    if r0str == r1str:
        state.regs.r0 = claripy.BVV(0x0, 8*4)
        print "Strings equal. R0 set to: ", state.se.any_int(state.regs.r0)
    else:
        state.regs.r0 = claripy.BVV(0x1, 8*4)
        print "Strings not equal. R0 set to: ", state.se.any_int(state.regs.r0)

b = angr.Project("./led-kmod.ko")
'''
The mutexes are irrelevant to us so we hook the call to the function 
to do nothing. I'm not hooking the function itself since I haven't
figured out how to do that. Instead, I pass "length=4" which instructs
angr to return to 4 bytes after the hooked code, which is the length
of the BL instruction.
'''
b.hook(0x400144, mutexlocknestedhook, length=4)
b.hook_symbol(u'strncmp', strncmphook)

'''
Let's call dev_write(int a1, char *status, char *outbuf)
'''
status = claripy.BVS('status', 8*7)
sym_out_buf = claripy.BVV(0x0, 8*4)

'''
The first and last arguments don't actually matter. The second is what
we need to solve for. What status value, when passed to this function, 
gets decoded to the string "FAILURE"?
'''
dev_write = b.loader.main_bin.get_symbol('dev_write').addr + b.entry
dev_write_state = b.factory.call_state(dev_write, 1, status, sym_out_buf)
pg = b.factory.path_group(dev_write_state, immutable=False)
pg.explore(find=0x40016c).unstash(from_stash='found',to_stash='active')

'''
Since this entire function is dependent on these symbolic values we're
passing in, we can just set constraints such that XOR(byte,0x39) for each
byte gives us our FAILURE status.
'''
s = dev_write_state
decodedptr = s.regs.sp.args[0]
encodedptr = s.regs.sp.args[0]+0x17
msg = [0x46,0x41,0x49,0x4c,0x55,0x52,0x45]
for i in range(len(msg)):
    d = s.memory.load(decodedptr + i, 1)
    s.add_constraints(d == msg[i])
    x = s.memory.load(encodedptr + i, 1)
    s.add_constraints(x^0x39 == d)

pg = b.factory.path_group(dev_write_state, immutable=False)

'''
This is the top of the loop in dev_write that does some bit-twiddling
'''
pg.explore(find=0x400154).unstash(from_stash='found',to_stash='active')
    
'''
Finish the loop
'''
pg.explore(find=0x40016c).unstash(from_stash='found',to_stash='active')
'''
We can view 64 bytes of the stack with this: 
s = pg.active[0].state
s.se.any_str(s.memory.load(s.regs.sp.args[0], 64))
'''
s = pg.active[0].state
enc = s.se.any_str(s.memory.load(s.regs.sp.args[0]+1, 7))
print "%s results in '%s'" % (repr(enc), s.se.any_str(s.memory.load(decodedptr, 7)))

'''
Now that we've found the correct status value, let's start over and pass
that in for the second parameter.
'''
dev_write_state = b.factory.call_state(0x40012c, 1, enc, sym_out_buf)
pg = b.factory.path_group(dev_write_state, immutable=False)

'''
All the strncmp calls...
'''
b.hook_symbol('strncmp', strncmphook)

#b.hook(0x400178, strncmphook, length=4)
#b.hook(0x400190, strncmphook, length=4)
#b.hook(0x4001e8, strncmphook, length=4)
#b.hook(0x400200, strncmphook, length=4)
#b.hook(0x400218, strncmphook, length=4)

'''
Get through the other statuses while avoiding the extra paths we don't 
care about.
'''
pg.explore(find=0x400184, avoid=0x40023c).unstash(from_stash='found',to_stash='active')
pg.explore(find=0x4001dc, avoid=0x4001a0).unstash(from_stash='found',to_stash='active')
pg.explore(find=0x4001f4, avoid=0x400244).unstash(from_stash='found',to_stash='active')
pg.explore(find=0x40020c, avoid=0x40026c).unstash(from_stash='found',to_stash='active')
pg.explore(find=0x400224, avoid=0x400280).unstash(from_stash='found',to_stash='active')
print "\n-------------------------------------------------------------------------------------------"
print "Found a path to ACMELEDFailure on a 'FAILURE' status from %s!." % repr(enc)
print "-------------------------------------------------------------------------------------------"
```

P.S. - Definitely looking forward to PR #544. I've had some trouble wrapping my head around Simuvex vs Angr.

**Comment by zardus** [comment]
Hey! This would make a great example -- I really hope they allow distribution!

We just got back from DEFCON, so I haven't had a chance to look into this yet, but now that #544 is merged (which also had some analogous changes in CLE), could you check to see if that issue still persists? Then we'll go from there :-)

**Comment by rhelmot** [comment]
My initial guess is that your libraries use relocations that are unsupported by angr. Do you get any warning messages while loading your project? If you don't and are worried that you aren't configured to see warnings, an easy way to check if you are is to call project.factory.{path_group,simgr}.run(). This should immediately print out a warning.

**Comment by sraboy** [comment]
@zardus, still an issue after updating to 58d0120 (and pulling in my changes in PR #554). I left it running for about an hour before killing it.

@rhelmot, here's what I get when loading the project with DEBUG logging enabled:
    
```
In [1]: import angr, claripy, logging
   ...: logging.getLogger('angr').setLevel(logging.DEBUG)
   ...: 
   ...: b = angr.Project("led-kmod.ko")
   ...: 
WARNING | 2017-08-02 12:17:47,697 | claripy | Claripy is setting the recursion limit to 15000. If Python segfaults, I am sorry.
INFO    | 2017-08-02 12:17:48,128 | angr.project | Loading binary led-kmod.ko
WARNING | 2017-08-02 12:17:48,131 | cle.relocations | Unknown reloc 28 on ARMEL
WARNING | 2017-08-02 12:17:48,132 | cle.relocations | Unknown reloc 29 on ARMEL
WARNING | 2017-08-02 12:17:48,137 | cle.relocations | Unknown reloc 42 on ARMEL
WARNING | 2017-08-02 12:17:48,148 | cle.loader | The main binary is a position-independent executable. It is being loaded with a base address of 0x400000.
DEBUG   | 2017-08-02 12:17:48,150 | angr.project | hooking 0x1000000 with <class 'angr.simos.CallReturn'>
DEBUG   | 2017-08-02 12:17:48,150 | angr.project | hooking 0x1000010 with <class 'angr.simos.LinuxLoader'>
DEBUG   | 2017-08-02 12:17:48,150 | angr.project | hooking 0x1000050 with <Hook for LinuxLoader (1 arg) (continuation)>
DEBUG   | 2017-08-02 12:17:48,150 | angr.project | hooking 0x1000020 with <class 'angr.simos._dl_rtld_lock_recursive'>
DEBUG   | 2017-08-02 12:17:48,150 | angr.project | hooking 0x1000030 with <class 'angr.simos._dl_rtld_unlock_recursive'>
DEBUG   | 2017-08-02 12:17:48,150 | angr.project | hooking 0x1000040 with <class 'angr.simos._vsyscall'>

In [2]: b.factory.path_group().run()
WARNING | 2017-08-02 12:46:54,511 | angr.path_group | No completion state defined for path group; stepping until all paths deadend
DEBUG   | 2017-08-02 12:46:54,513 | angr.path_group | Round 0: stepping <PathGroup with 1 active>
WARNING | 2017-08-02 12:46:54,826 | simuvex.engine.successors | Exit state has over 257 possible solutions. Likely unconstrained; skipping. <BV32 reg_40_1_32{UNINITIALIZED}>
INFO    | 2017-08-02 12:46:54,826 | angr.path_history | Callstack on the path is unbalanced.
DEBUG   | 2017-08-02 12:46:54,827 | angr.path_group | Out of paths in stash active
```
Which makes sense since I should set an entry point. When I actually set up the the `call_state` first, like above, I get the same warning then it appears to step forever:
```
In [6]: pg.run()
WARNING | 2017-08-02 12:50:44,657 | angr.path_group | No completion state defined for path group; stepping until all paths deadend
DEBUG   | 2017-08-02 12:50:44,658 | angr.path_group | Round 0: stepping <PathGroup with 1 active>
...
```
I'm going to look in to the relocation a bit. Any recommendations on where in the CLE to dig?

**Comment by rhelmot** [comment]
Yep, those are the culprit. You can find our relocation implementations in `cle/cle/backends/relocations`. Most of them are generic data relocations, but you can implement your own relocation class easily. Look at the Relocation base class, you can override any of its methods you want in order to make the operation smooth.

**Comment by sraboy** [comment]
Not sure if the pip packages recently changed but once I started working implementing `R_ARM_CALL` and made changes to cle, I reinstalled it and started getting an error. To make sure I didn't screw something up, I wiped my pip cache and virtualenvs, re-cloned angr-dev and reinstalled to a new virtualenv.

Here's what I got:
```
In [1]: import logging
   ...: import angr, claripy
   ...: logging.getLogger('angr').setLevel(logging.DEBUG)
   ...: logging.getLogger('cle').setLevel(logging.DEBUG)
   ...: b = angr.Project("led-kmod.ko")
   ...: 
WARNING | 2017-08-02 16:38:13,407 | claripy | Claripy is setting the recursion limit to 15000. If Python segfaults, I am sorry.
INFO    | 2017-08-02 16:38:13,839 | angr.project | Loading binary led-kmod.ko
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-f77737ea9ef8> in <module>()
      3 logging.getLogger('angr').setLevel(logging.DEBUG)
      4 logging.getLogger('cle').setLevel(logging.DEBUG)
----> 5 b = angr.Project("led-kmod.ko")

/home/slavoie/code/angr-dev/angr/angr/project.pyc in __init__(self, thing, default_analysis_mode, ignore_functions, use_sim_procedures, exclude_sim_procedures_func, exclude_sim_procedures_list, arch, simos, load_options, translation_cache, support_selfmodifying_code, **kwargs)
    140             l.info("Loading binary %s", thing)
    141             self.filename = thing
--> 142             self.loader = cle.Loader(self.filename, **load_options)
    143 
    144         # Step 2: determine its CPU architecture, ideally falling back to CLE's guess

/home/slavoie/code/angr-dev/cle/cle/loader.pyc in __init__(self, main_binary, auto_load_libs, force_load_libs, skip_libs, main_opts, lib_opts, custom_ld_path, ignore_import_version_numbers, rebase_granularity, except_missing_libs, gdb_map, gdb_fix, aslr, page_size)
    105         self.requested_objects = set()
    106         self.tls_object = None
--> 107         self._load_main_binary()
    108 
    109         if gdb_map is not None:

/home/slavoie/code/angr-dev/cle/cle/loader.pyc in _load_main_binary(self)
    181                                             if self._main_binary_stream is None
    182                                             else self._main_binary_stream,
--> 183                                         is_main_bin=True)
    184         self.memory = Clemory(self.main_bin.arch, root=True)
    185         self.add_object(self.main_bin)

/home/slavoie/code/angr-dev/cle/cle/loader.pyc in load_object(self, path, compatible_with, is_main_bin, backend, **kwargs)
    259 
    260         # do the load!!!
--> 261         return backend_cls(path, compatible_with=compatible_with, is_main_bin=is_main_bin, loader=self, **kwargs)
    262 
    263     @staticmethod

/home/slavoie/code/angr-dev/cle/cle/backends/elf.py in __init__(self, binary, **kwargs)
    219         seg_addrs = (ALIGN_DOWN(x['p_vaddr'], self.loader.page_size)
    220                      for x in self.reader.iter_segments() if x.header.p_type == 'PT_LOAD' and x.header.p_memsz > 0)
--> 221         self.mapped_base = self.linked_base = min(seg_addrs) if seg_addrs else 0
    222         self.__register_segments()
    223         self.__register_sections()

ValueError: min() arg is an empty sequence
```
Gonna dig through previous commits to see what might have broken that.

**Comment by rhelmot** [comment]
Oh, that's an actual bug. This is related to recent commits. You can see there was an attempt to avoid this crash in `min(seg_addrs) if seg_addrs else 0`, but seg_addrs is a generator, not an actual list, so it's always truthy. The best thing is probably to set those variables to zero, then try to take the min of the list and catch the ValueError.

**Comment by sraboy** [comment]
I just noticed it. If this is what you've got in mind, I'll do a PR:
```
seg_addrs = (ALIGN_DOWN(x['p_vaddr'], self.loader.page_size)
             for x in self.reader.iter_segments() if x.header.p_type == 'PT_LOAD')
self.mapped_base = self.linked_base = 0
try:
    self.mapped_base = self.linked_base = min(seg_addrs)
except ValueError:
    l.warn('no segments identified in PT_LOAD')
```

**Comment by rhelmot** [comment]
Excellent. Thanks!

**Comment by sraboy** [comment]
Okay, so I've got the basics of `R_ARM_CALL` relocation implemented. I've got the below class in `generic.py` and `R_ARM_CALL = generic.ARMPC24RelativeAddendReloc` in `arm.py`.

```
class ARMPC24RelativeAddendReloc(Relocation):
    @property
    def value(self):
        # (S + A) | T - P
        S = self.resolvedby.rebased_addr
        A = self.addend
        P = self.rebased_addr
        tmp = P
        l.debug("(S + A) | T - P")
        l.debug("S = self.resolvedby.rebased_addr = %s", hex(self.resolvedby.rebased_addr))
        l.debug("A = self.addend = %s", hex(self.addend))
        l.debug("P = self.rebased_addr = %s", hex(self.rebased_addr))
        l.debug("tmp = P = %s" % hex(tmp))
        if tmp & 0x00800000:
            tmp |= 0xff000000
            l.debug("sign extended") 
            l.debug("tmp |= |= 0xff000000 = %s" % hex(tmp))
        tmp = S + (tmp << 2) - P
        l.debug("tmp = S + (tmp << 2) - P = %s" % hex(tmp))
        tmp >>= 2
        l.debug("tmp >>=2 = %s" % hex(tmp))
        P = (P & 0xff000000) | (tmp & 0x00ffffff)
        l.debug("P = (P & 0xff000000) | (tmp & 0x00ffffff) = %s" % hex(tmp))
        l.debug("%s relocated to: %s", self.symbol.name, hex(P))
        return P
```

This appears to work great and I can hook any `R_ARM_CALL` symbol fine. Problem is that I've only got this one binary to test with that I can't share (at least not yet... I've asked).

Also, this should cover `R_ARM_PLT32`, `R_ARM_PC24`, and `R_ARM_JUMP24`. I can only verify `R_ARM_JUMP24` for now.

However, I think there may be a few kinks:
```
In [6]: x=b.loader.main_bin.get_symbol('strncmp')
   ...: print "%s is at %s" % (x.name, hex(x.addr))
   ...: x=b.loader.main_bin.get_symbol('mutex_lock_nested')
   ...: print "%s is at %s" % (x.name, hex(x.addr))
   ...: x=b.loader.main_bin.get_symbol('printk')
   ...: print "%s is at %s" % (x.name, hex(x.addr))
   ...: 
strncmp is at 0x0
mutex_lock_nested is at 0x0
printk is at 0x0
In [7]: for sym in b.loader.main_bin.symbols_by_addr:
   ...:     o = b.loader.main_bin.symbols_by_addr[sym]
   ...:     print "%s: %s" % (hex(o.addr+b.entry), o.name)
   ...:     
0x400408: $d
0x400010: FlipData
0x400498: $d
0x400768: 
0x400038: ACMELEDSuccess
0x400298: $d
0x40051c: __UNIQUE_ID_author1
0x4004a4: 
0x4000a8: ACMELEDFailure
0x400cdc: acmeled_mutex
0x40012c: dev_write
0x400530: __UNIQUE_ID_license0
0x400488: $d
0x40053c: __UNIQUE_ID_srcversion1
0x400eb8: pinON
0x4002bc: LED_Kmod_Init
0x400ec0: __key.23592
0x400ec8: acmeLEDCharClass
0x400ecc: acmeLEDCharDevice
0x400ed0: __key.23594
0x400438: LED_Kmod_Exit
0x4004d4: 
0x400d24: 
0x4004dc: 
0x400560: __module_depends
0x4004e4: 
0x400c68: fops
0x400ebc: majorNumber
0x4004f0: __UNIQUE_ID_description2
0x4005a8: $d
0x400008: dev_release
0x400569: __UNIQUE_ID_vermagic0
0x400e94: 
In [8]: b.loader.main_bin.imports
Out[8]: 
{u'gpiod_set_raw_value': <cle.backends.relocations.generic.ARMPC24RelativeAddendReloc at 0x7fbb48dea050>,
 u'printk': <cle.backends.relocations.generic.ARMPC24RelativeAddendReloc at 0x7fbb48deab90>}
```

I understand why `get_symbol` returns `0x0`... I just don't like it. For the second part, what's with the `$d` and where are these symbols? Last, why does `printk` show up but the not the others?

**Comment by rhelmot** [comment]
Considering that that relocation is only for ARM, you should put it in `arm.py`, not `generic.py` :)

Those symbols are at address 0 because they're the "import symbols", not the symbols that actually give an address for strcmp etc. To get the corresponding "export symbol" you can say `symbol.resolvedby`.

The $d etc symbols are probably just a weird compiler artifact. You can compare it against the ground truth with `readelf -s` on the binary, iirc.

Finally, we won't actually pick up the symbol as an import unless we can process a relocation using it.

**Comment by sraboy** [comment]
Though I can't provide tests yet, I'm cleaning this up (and moving it to `arm.py`) so it's easier to add some other similar relocations (eventually). I'm not holding my breath on being able to release the binaries so I'm not sure I'll have any test cases available and don't have any other ARM stuff around.

Once I get it all done and tested locally, I'll do a PR so it'll at least run against the CI and get some feedback.

You were right on the `$d` symbols too. I just hadn't noticed them before.

**Comment by sraboy** [comment]
Okay, so until about 2mos ago, my knowledge of ARM began with my Android phone and ended with my Roku. I just ordered an RPi so I'm going to do some testing over the next week or two. I've implemented R_ARM_CALL, R_ARM_JUMP24, R_ARM_PC24 and R_ARM_PREL31. They work to the extent that, since I give CLE relocations, I can hook the symbols which now have relocation addresses. What I'm not sure of is whether the relocations actually match how the linker would do it.

Anywho, here's my diff: https://github.com/angr/cle/compare/master...sraboy:wip/armreloc.

Aside from the actual functionality, let me know if the formatting and style should change.
